web/server/codechecker_server/api/authentication.py

```diff
@@ -167,6 +206,85 @@ def performLogin(self, auth_method, auth_string):
                     codechecker_api_shared.ttypes.ErrorCode.AUTH_DENIED,
                     msg)
 
+        elif auth_method == "oauth":
+            provider, url = auth_string.split("@")
+
+            oauth_config = self.__manager.get_oauth_config(provider)
+            if not oauth_config.get('enabled'):
+                LOG.error("OAuth authentication is " +
+                          "not enabled for provider: %s", provider)
+                raise codechecker_api_shared.ttypes.RequestFailed(
+                    codechecker_api_shared.ttypes.ErrorCode.AUTH_DENIED,
+                    "OAuth authentication is not enabled.")
+
+            client_id = oauth_config["oauth_client_id"]
+            client_secret = oauth_config["oauth_client_secret"]
+            scope = oauth_config["oauth_scope"]
+            token_url = oauth_config["oauth_token_uri"]
+            user_info_url = oauth_config["oauth_user_info_uri"]
+            redirect_uri = oauth_config["oauth_redirect_uri"]
+            allowed_users = oauth_config.get("allowed_users", [])
+
+            LOG.info("OAuth configuration loaded for provider: %s", provider)
+
+            session = OAuth2Session(
+                client_id,
+                client_secret,
+                scope=scope,
+                redirect_uri=redirect_uri)
+
+            # FIXME: This is a workaround for the Microsoft OAuth2 provider
+            # which doesn't correctly fetch the code from url.
+
+            url_new = urlparse(url)
+            parsed_query = parse_qs(url_new.query)
+
+            code = parsed_query.get("code")[0]
+            state = parsed_query.get("state")[0]
+
+            url = url_new.scheme + "://" + url_new.netloc + url_new.path + \
+                "?code=" + code + "&state=" + state
```

Comment:
Not an issue here, but take care that `parse_qs` does a URL decode on the query parameters: `parse_qs(urlparse("https://aaa.aaa/?value=%3F%23%3D%26").query) => {'value': ['?#=&']}`

Reply:
Understood